hare/eligibility/oracle.go

```diff
@@ -388,167 +373,93 @@ func (o *Oracle) Proof(ctx context.Context, nonce types.VRFPostIndex, layer type
 
 // Returns a map of all active node IDs in the specified layer id.
 func (o *Oracle) actives(ctx context.Context, targetLayer types.LayerID) (map[types.NodeID]uint64, error) {
+	if !targetLayer.After(types.GetEffectiveGenesis()) {
+		return nil, errEmptyActiveSet
+	}
+	targetEpoch := targetLayer.GetEpoch()
+	firstEpoch := types.GetEffectiveGenesis().GetEpoch() + 1
+	// the first bootstrap data targets first epoch after genesis (epoch 2)
+	if targetEpoch != firstEpoch &&
+		targetLayer.Difference(targetEpoch.FirstLayer()) < o.cfg.ConfidenceParam {
+		targetEpoch -= 1
+	}
 	logger := o.WithContext(ctx).WithFields(
 		log.FieldNamed("target_layer", targetLayer),
 		log.FieldNamed("target_layer_epoch", targetLayer.GetEpoch()),
+		log.FieldNamed("target_epoch", targetEpoch),
 	)
 	logger.Debug("hare oracle getting active set")
 
 	o.lock.Lock()
 	defer o.lock.Unlock()
-	// we first try to get the hare active set for a range of safe layers
-	safeLayerStart, safeLayerEnd := safeLayerRange(
-		targetLayer, o.cfg.ConfidenceParam, o.layersPerEpoch, o.cfg.EpochOffset)
-	logger.With().Debug("safe layer range",
-		log.FieldNamed("safe_layer_start", safeLayerStart),
-		log.FieldNamed("safe_layer_end", safeLayerEnd),
-		log.FieldNamed("safe_layer_start_epoch", safeLayerStart.GetEpoch()),
-		log.FieldNamed("safe_layer_end_epoch", safeLayerEnd.GetEpoch()),
-		log.Uint32("confidence_param", o.cfg.ConfidenceParam),
-		log.Uint32("epoch_offset", o.cfg.EpochOffset),
-		log.Uint64("layers_per_epoch", uint64(o.layersPerEpoch)),
-		log.FieldNamed("effective_genesis", types.GetEffectiveGenesis()))
-
-	if value, exists := o.activesCache.Get(safeLayerStart.GetEpoch()); exists {
+
+	// note: for simplicity, bootstrap active set will be used btwn
+	// [types.EpochID(2).FirstLayer(), types.EpochID(3).Add(o.cfg.ConfidenceParam)-1]
+	// because the first bootstrap data targets epoch 2 (atxs were published in epoch 1)
```

Comment:
@dshulyak 
i know this differ from our previous discussion. the reason being that each update published by spacemesh is designated with a target epoch (for beacon and activeset)
so for epoch 2, oracle will receive an update for epoch 2 as the bootstrap value and use it for the first `confidence_param` layers and cache this value as epoch 2 cache.

i didn't want to change the cached value after `confidence_param` has passed, as it's literally the same active set.

Reply:
at the beginning of the network, what you described would be (`bs`: bootstrap, `bk10`: block generated in layer 10)

| layer     | 10 | 11 | 12 | 13 | 14 |
| ---      | :---: | :---: | :---: | :---: | :---: |
| block | bk10 | bk11 | bk12 | bk13 | bk14 |
| AS     | bs | bs | bk10 | bk10 | bk10 |

| layer     | 15 | 16 | 17 | 18 | 19 |
| ---      | :---: | :---: | :---: | :---: | :---: |
| block | bk15 | bk16 | bk17 | bk18 | bk19 |
| AS     | bk10 | bk10 | bk15 | bk15 | bk15 |

whereas my approach would be

| layer     | 10 | 11 | 12 | 13 | 14 |
| ---      | :---: | :---: | :---: | :---: | :---: |
| block | bk10 | bk11 | bk12 | bk13 | bk14 |
| AS     | bs | bs | bs | bs | bs |

| layer     | 15 | 16 | 17 | 18 | 19 |
| ---      | :---: | :---: | :---: | :---: | :---: |
| block | bk15 | bk16 | bk17 | bk18 | bk19 |
| AS     | bs | bs | bk15 | bk15 | bk15 |

and because the `bk10` is a subset of `bs`, since the cache is keyed by epoch, once saved, i didn't bother updating the cache after confidence param passed. but not a big deal. i can change that.

---

- how we handle empty layers

let's say this is epoch 3. in my current implementation, it would be (`fb3`: fallback for epoch 3)
| layer     | 15 | 16 | 17 | 18 | 19 |
| ---      | :---: | :---: | :---: | :---: | :---: |
| block | X | X | X | bk18 | bk19 |
| AS     | bk10 | bk10 | fb3 | fb3 | fb3 |

the question here is, should we use `bk18` for active set at layer 19? i.e.
| layer     | 15 | 16 | 17 | 18 | 19 |
| ---      | :---: | :---: | :---: | :---: | :---: |
| block | X | X | X | bk18 | bk19 |
| AS     | bk10 | bk10 | fb3 | fb3 | bk18 |

i didn't bother switching (same reason as above). my thought is that there is already consensus. why switch to a different/weaker one.

- no progress in previous epoch, such as
> [1, 1, 0, 0, 0]
[0, 0, ?]

in my impl, i don't really care about failures in previous epoch. the only criteria is that by confidence layer, if there is no block generated "for the current epoch", the fallback value is used and saved in cache for subsequent layers.

| layer     | 15 | 16 | 17 | 18 | 19 |
| ---      | :---: | :---: | :---: | :---: | :---: |
| block | X | X | X | X | X |
| AS     | bk10 | bk10 | fb3 | fb3 | fb3 |

| layer     | 20 | 21 | 22 | 23 | 24 |
| ---      | :---: | :---: | :---: | :---: | :---: |
| block | bk20 | bk21 | bk22 | bk23 | bk24 |
| AS     | fb3 | fb3 | bk20 | bk20 | bk20 |

- we see that nodes disagree about active set and we want to help them

i have not thought of covering this case. here is my understanding of your description,
node A has
| layer     | 20 | 21 | 22 | 23 | 24 |
| ---      | :---: | :---: | :---: | :---: | :---: |
| block | bk20A | bk21 | bk22 | bk23 | bk24 |
| AS     | bk15 | bk15 | bk20A | bk20A | bk20A |

and node B has
| layer     | 20 | 21 | 22 | 23 | 24 |
| ---      | :---: | :---: | :---: | :---: | :---: |
| block | bk20B | bk21 | bk22 | bk23 | bk24 |
| AS     | bk15 | bk15 | bk20B | bk20B | bk20B |

correct? 
so if this is network split, then we would have 2 set of hare protocol running in the network. Tal aslo stated something in the past here https://spacemesh.slack.com/archives/CG91F13C7/p1667953163180779

if not a network split, the contents of `bk20A` and `bk20B` should be the same (?) especially after the ATX grading is applied.

a similar problem arise here. say node B changed its mind and reverted to `bk20A` during layer 23. do you think node should switch to `bk20A` for active set in layer 24? i think yes.

my opinion is, once there is consensus, there is no need to change it to a different consensus.
so the last case (`bk20B`->`bk20A`) is the only case that i think worthy of changing.

Reply:
i think this discussion no longer applies.